### ISSUE TITLE
NBEthernet example code

### ISSUE THREAD

**Issue by rodjt** [DavyLandman/NBEthernet#1]
I am new to Github so I hope this is the correct place for asking the following...

My experience is mainly hardware design but I have had some experience with embedded 'C'.

I have an arduino field data monitoring project that requires the Ethernet library. The main controller (Meg2560) is handling all the field inputs and RS232 comms (terminal refreshes) and timing so I can not afford for the main loop to just halt while the Ethernet library functions impose delays. I am looking for exactly what you have written ie a 'non-blocking' Ethernet library.

I have copied and installed the library at

https://github.com/DavyLandman/NBEthernet

and copied your example code at

https://github.com/arduino/Arduino/pull/1240

I have added the library to the project which added the headers to the Ethernet_State_Machine.ino file

(all the 'left arrows' left out of the following so they display in the text)
# include util.h>
# include NBEthernet.h>
# include  EthernetUdp.h>
# include EthernetServer.h>
# include EthernetClient.h>
# include Dns.h>
# include Dhcp.h>

I am using Atmel Studio 6 with Visual micro. (Arduino 1.5..)
I get lots of compile errors such as

Ethernet_State_Machine.ino:18:14: error: 'class EthernetClass' has no member named 'initialize'

so I know I have something very basic very wrong.
Have I got the right version of the library?
Have I got the right example code?

Any help really appreciated. Thanks.

**Comment by DavyLandman** [comment]
Hi rodjt, 

I would advise reporting your interest in the pull-request (arduino/Arduino#1240) discussion.

However, sadly it hasn't been merged for 2 years, so there is very little chance of it ever happening.

The code must have evolved from then.

If you would want to run it, I would try the code from that pull request. This repo is outdated, since I switched from making a separate library to adding it to the arduino core (since it only required small changes.)

Who knows, I might spend some time extracting the code from the pull request and putting it back into this repo, but that has some binary size impact.. (and personal time)

**Comment by rodjt** [comment]
Hi Davy,

Thank you for replying so promptly.

In the past (typically commercial horticultural control projects) I have 
designed the hardware and written all the embedded code. (no library 
calls) That has given me total control and I have always written code so 
that the main loop continues with periferal functions being called when 
required via flags. Arduino was my attempt to avoid hardware 
development. There is so much that I like about it, but I had to upgrade 
the IDE and write code such that as much as possible I avoided the 
arduino libraries. I was amazed at how many libraries simply use delays 
instead of returning status conditions so that the main loop could 
continue. And also amazed at how little that seemed to matter to the 
community.

The current project is an agricultural monitor. Gathering data from 
across the farm and transmitting it to Xively via an onsite Ethernet 
connection. Because it is a 'real time' application I need the main loop 
to continue.

I have got the project almost complete having designed a hardware i/o 
board and used the arduino Ethernet lib. for communication. It is at the 
stage where everything is looking good write up to where I add the 
communication. The Ethernet connection works and delivers data, but also 
has large delays 5 seconds or so at times and after a variable number of 
transmissions completely hangs during a connect call and will not 
restart no matter what I do apart from a complete system reset. This 
makes it totally useless.

What you have started is exactly what I require and I would be prepared 
to put time into this to get it to a useable state but my experience in 
C++ is almost non existent and with code this complex I am not going to 
be able to make progress.

I will try to put together a working system from the code I see at the 
link you provided. I really appreciate the fact that you have replied 
and helped with direction. Any comments you can provide to guide me 
through the implementation would be really appreciated.

Just a comment. With respect to combining it with the arduino lib, from 
my perspective it would be better to leave it separate. The independence 
of the library is a good thing. If it makes it easier to complete as a 
separate library it would be really good to see that happen.

Thanks and regards
Rod

**Comment by DavyLandman** [comment]
Perhaps you should take a look at:
https://github.com/ntruchsess/arduino_uip

**Comment by rodjt** [comment]
Yes, thanks. I received a module for the ENC28J60 yesterday as a backup 
in case the W5100 didn't work out. I will check it out today.
